File: apps/desktop/src/lib/recovery.ts
// Recovery client for Tauri backend
// Handles crash recovery operations via IPC

import { invoke } from '@tauri-apps/api/core';

// ============================================================================
// Types (matching Rust types)
// ============================================================================

export interface RecoveryFile {
  file_key: string;
  wal_content: string;
  wal_time: string;
  workspace_root: string;
}

// Transformed type for frontend use (camelCase)
export interface RecoveryFileInfo {
  fileKey: string;
  walContent: string;
  walTime: Date;
  workspaceRoot: string;
}

// ============================================================================
// Recovery Client
// ============================================================================

class RecoveryClient {
  /**
   * Check for recovery files on startup
   * Returns list of files with unsaved changes
   */
  async checkForRecovery(workspaceRoot: string): Promise<RecoveryFileInfo[]> {
    const files = await invoke<RecoveryFile[]>('recovery_check', {
      workspaceRoot,
    });

    return files.map((f) => ({
      fileKey: f.file_key,
      walContent: f.wal_content,
      walTime: new Date(f.wal_time),
      workspaceRoot: f.workspace_root,
    }));
  }

  /**
   * Write WAL file for a document
   * Returns true if content was written (changed), false if skipped (unchanged)
   */
  async writeWal(workspaceRoot: string, fileKey: string, content: string): Promise<boolean> {
    return invoke<boolean>('recovery_write_wal', {
      workspaceRoot,
      fileKey,
      content,
    });
  }

  /**
   * Clear WAL file after successful save
   */
  async clearWal(workspaceRoot: string, fileKey: string): Promise<void> {
    await invoke('recovery_clear_wal', {
      workspaceRoot,
      fileKey,
    });
  }

  /**
   * Check if a specific file has recovery available
   */
  async hasRecovery(workspaceRoot: string, fileKey: string): Promise<boolean> {
    return invoke<boolean>('recovery_has_recovery', {
      workspaceRoot,
      fileKey,
    });
  }

  /**
   * Get recovery content for a specific file
   */
  async getRecoveryContent(workspaceRoot: string, fileKey: string): Promise<string | null> {
    return invoke<string | null>('recovery_get_content', {
      workspaceRoot,
      fileKey,
    });
  }

  /**
   * Discard recovery for a specific file (user chose not to recover)
   */
  async discardRecovery(workspaceRoot: string, fileKey: string): Promise<void> {
    await invoke('recovery_discard', {
      workspaceRoot,
      fileKey,
    });
  }

  /**
   * Discard all recovery files for a workspace
   */
  async discardAllRecovery(workspaceRoot: string): Promise<void> {
    await invoke('recovery_discard_all', {
      workspaceRoot,
    });
  }

  /**
   * Check if recovery content differs from current file content
   */
  async hasUniqueContent(
    workspaceRoot: string,
    fileKey: string,
    currentContent: string
  ): Promise<boolean> {
    return invoke<boolean>('recovery_has_unique_content', {
      workspaceRoot,
      fileKey,
      currentContent,
    });
  }
}

export const recoveryClient = new RecoveryClient();

File: apps/desktop/src-tauri/src/commands/recovery.rs
// Recovery commands - IPC handlers for crash recovery

use crate::services::recovery_manager::{RecoveryFile, RecoveryManager};
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::Arc;
use tauri::Runtime;
use tokio::sync::RwLock;
use tracing::{debug, info};

/// Registry of recovery managers (one per workspace)
pub struct RecoveryManagerRegistry {
    managers: HashMap<String, Arc<RecoveryManager>>,
}

impl RecoveryManagerRegistry {
    pub fn new() -> Self {
        Self {
            managers: HashMap::new(),
        }
    }

    pub async fn get_or_create(&mut self, workspace_root: &str) -> Arc<RecoveryManager> {
        if let Some(manager) = self.managers.get(workspace_root) {
            return manager.clone();
        }

        let manager = Arc::new(RecoveryManager::new(PathBuf::from(workspace_root)));

        // Initialize the recovery directory
        if let Err(e) = manager.init().await {
            tracing::warn!("Failed to initialize recovery manager: {}", e);
        }

        self.managers
            .insert(workspace_root.to_string(), manager.clone());
        manager
    }
}

impl Default for RecoveryManagerRegistry {
    fn default() -> Self {
        Self::new()
    }
}

/// State for recovery managers
pub struct RecoveryState {
    pub registry: RwLock<RecoveryManagerRegistry>,
}

impl RecoveryState {
    pub fn new() -> Self {
        Self {
            registry: RwLock::new(RecoveryManagerRegistry::new()),
        }
    }
}

impl Default for RecoveryState {
    fn default() -> Self {
        Self::new()
    }
}

// ============================================================================
// Tauri Commands
// ============================================================================

/// Check for recovery files on startup
/// Returns list of files with unsaved changes
#[tauri::command]
pub async fn recovery_check<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
) -> Result<Vec<RecoveryFile>, String> {
    debug!("Checking for recovery files in: {}", workspace_root);

    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    manager.check_for_recovery().await
}

/// Write WAL file for a document
/// Called periodically while editing to save recovery data
#[tauri::command]
pub async fn recovery_write_wal<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
    file_key: String,
    content: String,
) -> Result<bool, String> {
    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    manager.write_wal(&file_key, &content).await
}

/// Clear WAL file after successful save
#[tauri::command]
pub async fn recovery_clear_wal<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
    file_key: String,
) -> Result<(), String> {
    debug!("Clearing WAL for: {}", file_key);

    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    manager.clear_wal(&file_key).await
}

/// Check if a specific file has recovery available
#[tauri::command]
pub async fn recovery_has_recovery<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
    file_key: String,
) -> Result<bool, String> {
    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    Ok(manager.has_recovery(&file_key).await)
}

/// Get recovery content for a specific file
#[tauri::command]
pub async fn recovery_get_content<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
    file_key: String,
) -> Result<Option<String>, String> {
    debug!("Getting recovery content for: {}", file_key);

    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    manager.get_recovery_content(&file_key).await
}

/// Discard recovery for a specific file
#[tauri::command]
pub async fn recovery_discard<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
    file_key: String,
) -> Result<(), String> {
    info!("Discarding recovery for: {}", file_key);

    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    manager.discard_recovery(&file_key).await
}

/// Discard all recovery files for a workspace
#[tauri::command]
pub async fn recovery_discard_all<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
) -> Result<(), String> {
    info!("Discarding all recovery files in: {}", workspace_root);

    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    manager.discard_all_recovery().await
}

/// Check if recovery content differs from current file content
#[tauri::command]
pub async fn recovery_has_unique_content<R: Runtime>(
    _app: tauri::AppHandle<R>,
    state: tauri::State<'_, RecoveryState>,
    workspace_root: String,
    file_key: String,
    current_content: String,
) -> Result<bool, String> {
    let mut registry = state.registry.write().await;
    let manager = registry.get_or_create(&workspace_root).await;

    manager
        .has_unique_recovery(&file_key, &current_content)
        .await
}
